horreum-backend/src/test/java/io/hyperfoil/tools/horreum/svc/SchemaServiceNoRestTest.java

```diff
@@ -0,0 +1,639 @@
+      assertEquals(String.format("Label id=%d, name=%s belongs to a different schema: %d(%s)",
+            l.id, l.name, s.id, s.uri), thrown.getMessage());
+   }
+
+   @org.junit.jupiter.api.Test
+   void testUpdateSchemaLabelNameWhenAlreadyExisting() {
+      Schema s = createSchema("dummy", "urn:dummy:schema");
+
+      Label l = createSampleLabel("Blabla", s, null, "({x, y}) => ({ z: 1 })",
+            new Extractor("x", "$.x", true), new Extractor("y", "$.y", false));
+      Label another = createSampleLabel("AnotherLabel", s, null, "({x, y}) => ({ z: 1 })");
+
+      int id = schemaService.addOrUpdateLabel(s.id, l);
+      schemaService.addOrUpdateLabel(s.id, another);
+      assertEquals(2, LabelDAO.count());
+
+      // update the label
+      l.id = id;
+      l.name = "AnotherLabel";
+      ServiceException thrown = assertThrows(ServiceException.class, () -> schemaService.addOrUpdateLabel(s.id, l));
+      assertEquals(String.format("There is an existing label with the same name (%s) in this " +
+            "schema; please choose different name.", l.name), thrown.getMessage());
+   }
+
+   @org.junit.jupiter.api.Test
+   void testDeleteSchemaLabel() {
+      Schema s = createSchema("dummy", "urn:dummy:schema");
+
+      Label l = createSampleLabel("Blabla", s, null, "({x, y}) => ({ z: 1 })",
+            new Extractor("x", "$.x", true), new Extractor("y", "$.y", false));
+
+      int id = schemaService.addOrUpdateLabel(s.id, l);
+      assertEquals(1, LabelDAO.count());
+
+      schemaService.deleteLabel(s.id, id);
+      assertEquals(0, LabelDAO.count());
+   }
+
+   @org.junit.jupiter.api.Test
+   void testDeleteSchemaLabelWithFailures() {
+      Schema s = createSchema("dummy", "urn:dummy:schema");
+
+      Label l = createSampleLabel("Blabla", s, null, "({x, y}) => ({ z: 1 })",
+            new Extractor("x", "$.x", true), new Extractor("y", "$.y", false));
+
+      int id = schemaService.addOrUpdateLabel(s.id, l);
+      assertEquals(1, LabelDAO.count());
+
+      ServiceException thrown = assertThrows(ServiceException.class, () -> schemaService.deleteLabel(s.id, 999));
+      assertEquals(String.format("Label %d not found", 999), thrown.getMessage());
+
+      thrown = assertThrows(ServiceException.class, () -> schemaService.deleteLabel(999, id));
+      assertEquals(String.format("Label %d does not belong to schema %d", id, 999), thrown.getMessage());
+   }
+
+   // utility to create a schema in the db, tested with testCreateSchema
+   private Schema createSchema(String name, String uri) {
+      return createSchema(name, uri, null);
+   }
+
+   // utility to create a schema in the db, tested with testCreateSchema
+   private Schema createSchema(String name, String uri, String token) {
+      Schema schema = new Schema();
+      schema.owner = FOO_TEAM;
+      schema.name = name;
+      schema.uri = uri;
+      schema.token = token;
+      schema.id = schemaService.add(schema);
+      return schema;
+   }
```

Comment:
What is the difference between this and `BaseServiceNoRestTest.createSampleSchema()`?
Instead of duplicating code, we should use the method in the parent class

Reply:
The only difference is line `636`, `createSampleSchema` simply creates the DTO object.. whereas `createSchema` creates a DTO and call the service to add it to the db.

I am refactoring this to make use of `createSampleSchema` inside `createSchema` body, something like:
```java
   // utility to create a schema in the db, tested with testCreateSchema
   private Schema createSchema(String name, String uri, String token) {
      Schema schema = createSampleSchema(name, uri, FOO_TEAM, token);
      schema.id = schemaService.add(schema);
      return schema;
   }
```